glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift

```diff
@@ -0,0 +1,213 @@
+            // engines through the `Dispatchers.API` context, so this ensures we are enqueued
+            // before any other recording API call.
+            //
+            // * Do not change `Dispatchers.shared.serialOperationQueue.addOperation` to
+            // `Dispatchers.shared.launchAPI` as this would break startup overdue ping
+            //  collection. * `addOperation` schedules the task for immediate execution on the
+            // `Dispatchers` serial execution queue, before any other enqueued task. For more
+            // context, see bug 1604861 and the implementation of
+            // `collectPingAndReschedule`.
+            Dispatchers.shared.serialOperationQueue.addOperation {
+                self.collectPingAndReschedule(now, startupPing: true)
+            }
+        } else {
+            // This covers (3).
+            logger.info("The 'metrics' collection is scheduled for today, \(now)")
+            schedulePingCollection(now, sendTheNextCalendarDay: false)
+        }
+    }
+
+    /// Triggers the collection of the "metrics" ping and schedules the next collection.
+    ///
+    /// - parameters:
+    ///     * now: A `Date` representing the current time
+    @objc func collectPingAndReschedule(_ now: Date, startupPing: Bool = false) {
+        logger.info("Collecting the 'metrics' ping, now = \(now), startup = \(startupPing)")
+        if startupPing {
+            // **IMPORTANT**
+            //
+            // During the Glean initialization, we require any metric recording to be
+            // batched up and replayed after any startup metrics ping is sent. To guarantee
+            // that, we dispatch this function from `Dispatchers.API.executeTask`. However,
+            // Pings.metrics.submit() ends up calling `Dispatchers.API.launch` again which
+            // will delay the ping collection task after any pending metric recording is
+            // executed, breaking the 'metrics' ping promise of sending a startup 'metrics'
+            // ping only containing data from the previous session.
+            // To prevent that, we synchronously manually dispatch the 'metrics' ping, without
+            // going through our public API.
+            //
+            // * Do not change this line without checking what it implies for the above wall
+            // of text. *
+            Glean.shared.submitPingsByNameSync(pingNames: ["metrics"])
+        } else {
+            GleanMetrics.Pings.shared.metrics.submit()
+        }
+        // Update the collection date: we don't really care if we have data or not, let's
+        // always update the sent date.
+        updateSentDate(now)
+        // Reschedule the collection.
+        schedulePingCollection(now, sendTheNextCalendarDay: true)
+    }
+
+    /// Get the date the metrics ping was last collected.
+    ///
+    /// - returns: An Optional `Date` representing the date the metrics ping was last collected, or null if no metrics
```

Comment:
```suggestion
    /// - returns: A `Date` representing the date the metrics ping was last collected, or nil if no metrics
```

IMo the optional is implied by the "or nil"